src/System.Runtime.Extensions/src/System/IO/Path.cs

```diff
@@ -0,0 +1,1345 @@
+        {
+            // This routine is optimised to be used with buffs of length 20
+            Contract.Assert(((buff.Length % 5) == 0), "Unexpected hash length");
+
+            StringBuilder sb = StringBuilderCache.Acquire();
+            byte b0, b1, b2, b3, b4;
+            int l, i;
+
+            l = buff.Length;
+            i = 0;
+
+            // Create l chars using the last 5 bits of each byte.  
+            // Consume 3 MSB bits 5 bytes at a time.
+
+            do
+            {
+                b0 = (i < l) ? buff[i++] : (byte)0;
+                b1 = (i < l) ? buff[i++] : (byte)0;
+                b2 = (i < l) ? buff[i++] : (byte)0;
+                b3 = (i < l) ? buff[i++] : (byte)0;
+                b4 = (i < l) ? buff[i++] : (byte)0;
+
+                // Consume the 5 Least significant bits of each byte
+                sb.Append(s_Base32Char[b0 & 0x1F]);
+                sb.Append(s_Base32Char[b1 & 0x1F]);
+                sb.Append(s_Base32Char[b2 & 0x1F]);
+                sb.Append(s_Base32Char[b3 & 0x1F]);
+                sb.Append(s_Base32Char[b4 & 0x1F]);
+
+                // Consume 3 MSB of b0, b1, MSB bits 6, 7 of b3, b4
+                sb.Append(s_Base32Char[(
+                        ((b0 & 0xE0) >> 5) |
+                        ((b3 & 0x60) >> 2))]);
+
+                sb.Append(s_Base32Char[(
+                        ((b1 & 0xE0) >> 5) |
+                        ((b4 & 0x60) >> 2))]);
+
+                // Consume 3 MSB bits of b2, 1 MSB bit of b3, b4
+
+                b2 >>= 5;
+
+                Contract.Assert(((b2 & 0xF8) == 0), "Unexpected set bits");
+
+                if ((b3 & 0x80) != 0)
+                    b2 |= 0x08;
+                if ((b4 & 0x80) != 0)
+                    b2 |= 0x10;
+
+                sb.Append(s_Base32Char[b2]);
+            } while (i < l);
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        // ".." can only be used if it is specified as a part of a valid File/Directory name. We disallow
+        //  the user being able to use it to move up directories. Here are some examples eg 
+        //    Valid: a..b  abc..d
+        //    Invalid: ..ab   ab..  ..   abc..d\abc..
+        //
+        internal static void CheckSearchPattern(String searchPattern)
+        {
+            int index;
+            while ((index = searchPattern.IndexOf("..", StringComparison.Ordinal)) != -1)
+            {
+                if (index + 2 == searchPattern.Length) // Terminal ".." . Files names cannot end in ".."
+                    throw new ArgumentException(SR.Arg_InvalidSearchPattern);
+
+                if ((searchPattern[index + 2] == DirectorySeparatorChar)
+                   || (searchPattern[index + 2] == AltDirectorySeparatorChar))
+                    throw new ArgumentException(SR.Arg_InvalidSearchPattern);
+
+                searchPattern = searchPattern.Substring(index + 2);
+            }
```

Comment:
This function has some unnecessary allocations; I previously fixed them in a copy of this function that lives in System.IO.FileSystem's PathHelpers... we should fix them here, too.
https://github.com/dotnet/corefx/commit/196d67401fe20e92fb7ce48efb9de8fcdb3e5d9d